department-of-veterans-affairs/vets-api
# 60787 anti spam controls

## Summary

This PR further expands upon our initial work of creating a temporary service to send out login credential adoption related emails to eligible veteran users.  The main goal of this PR is to add functionality to prevent us from repeatedly sending emails to veterans within a one (1) week time frame since their last login that triggered an email send.

Put slightly more plainly:  While users may log in multiple times within a one (1) week period, with the work in this PR they should only receive one of these login action triggered Credential Adoption emails once every seven (7) days.

In specific, this PR makes the following changes:
- adds additional methods in the AcceptableVerifiedCredentialAdoptionService to create and query the new database table to record and then later evaluate if a specific user has previously triggered an email send within the last seven (7) days.
- updates/refactors some of the tests
- updates/refactors some of the naming conventions used in the service and tests
- adds a factory for test usage
- adds a new model (with validations) in accordance with the aforementioned new database table

The basic flow of this solution is as follows:
- once called, the AcceptableVerifiedCredentialAdoptionService evaluates if the veteran user has the appropriate credentials to be considered eligible to receive the Credential Adoption email.  This evaluation consists of:
  - checking if the user has both a DSLogon (legacy) and Login.gov (modern) credential
  - checking if the user's most recent login was with the legacy or modern credential 
  - checking if there is a record in the new database table with an email_triggered_at timestamp within the last seven (7) days
- If the user meets the eligibility criteria, we process the message for sending via VA Notify and then capture the appropriate events:
  - two (2) email send action events logged via StatsD
  - one (1) new record created in the CredentialAdoptionEmailRecords table
- If the user does not meet the first two (2) evaluation criteria (possessing both credentials and using the legacy credential most recently), the flow stops and nothing more happens.

_Please note:  The code changes required to call this service via the AfterLoginActions service will be made in a separate PR after this one is fully reviewed, approved, and merged in to main/master._

This work is being done by the Login.gov Adoption Team which reports to John Rahaghi, Product Owner with OCTO on the VA side.


## Related issue(s)
- [ZenHub Ticket 60787](https://app.zenhub.com/workspaces/logingov-adoption-team-632280b31e745000136a61fc/issues/gh/department-of-veterans-affairs/va.gov-team/60787)


## Testing done

Prior Behavior:
The prior behavior for this feature was that if called, it would only determine the first two (2) eligibility criteria and, if those are met, send the email regardless of how many times it may have been triggered and sent before.

Steps to Verify:
Since this functionality is not currently hooked up to the after_login_actions service, there is no user facing way to directly trigger the AcceptableCredentialAdoptionService perform method.  If an engineer were to use a rails console to manually call this service and pass in a user, that user should be evaluated as expected depending on what eligibility criteria they meet/don't meet.  

## What areas of the site does it impact?
Currently it impacts no user facing portions of the site.  On the non-user facing side, it only makes changes to currently unused backend code and adds a new database table which is not in use by or tied to any other feature/service.

## Acceptance criteria

- [x]  I added and updated unit tests and integration tests for each feature (if applicable).
